cmd/ping.go

```diff
@@ -264,12 +193,105 @@ func fetchAdminInfo(admClnt *madmin.AdminClient) (madmin.InfoMessage, error) {
 			if e == nil {
 				return info, nil
 			}
-
 			timer.Reset(time.Second)
 		}
 	}
 }
 
+func ping(ctx context.Context, cliCtx *cli.Context, anonClient *madmin.AnonymousClient, admInfo madmin.InfoMessage, endPointMap map[string]serverStats, index int) {
+	var endPointStats []EndPointStats
+	var servers []madmin.ServerProperties
+	if cliCtx.Bool("distributed") {
+		servers = admInfo.Servers
+	} else {
+		servers = append(servers, admInfo.Servers[0])
+	}
+
+	for result := range anonClient.Alive(ctx, madmin.AliveOpts{}, servers...) {
+		endPoint := getEndPoint(result)
+		stat := populateData(cliCtx, result, endPointMap)
+		endPointStat := EndPointStats{
+			Endpoint:  endPoint,
+			Min:       time.Duration(stat.min).Round(time.Microsecond).String(),
+			Max:       time.Duration(stat.max).Round(time.Microsecond).String(),
+			Average:   time.Duration(stat.avg).Round(time.Microsecond).String(),
+			CountErr:  strconv.Itoa(stat.errorCount),
+			Error:     stat.err,
+			Roundtrip: result.ResponseTime.Round(time.Microsecond).String(),
+		}
+		endPointStats = append(endPointStats, endPointStat)
+		endPointMap[result.Endpoint.Host] = stat
+
+	}
+	printMsg(PingResult{
+		Status:    "success",
+		Counter:   strconv.Itoa(index),
+		Endpoints: endPointStats,
+	})
+
+	time.Sleep(time.Duration(cliCtx.Int("interval")) * pingInterval)
+}
+
+func populateData(cliCtx *cli.Context, result madmin.AliveResult, serverMap map[string]serverStats) serverStats {
+	ec := cliCtx.Int("error-count")
+	var errorString string
+	var sum, avg uint64
+	min := uint64(math.MaxUint64)
+	var max uint64
+	var counter, errorCount int
+
+	if result.Error != nil {
+		errorString = result.Error.Error()
+		if stat, ok := serverMap[result.Endpoint.Host]; ok {
+			min = stat.min
+			max = stat.max
+			sum = stat.sum
+			counter = stat.counter
+			avg = stat.avg
+			errorCount = stat.errorCount + 1
+
+		} else {
+			min = 0
+			errorCount = 1
+		}
+		if errorCount >= ec {
+			stop = true
+		}
+	} else {
+		// reset consecutive error count
+		errorCount = 0
+		if stat, ok := serverMap[result.Endpoint.Host]; ok {
+			var minVal uint64
+			if stat.min == 0 {
+				minVal = uint64(result.ResponseTime)
+			} else {
+				minVal = stat.min
+			}
+			min = uint64(math.Min(float64(minVal), float64(uint64(result.ResponseTime))))
+			max = uint64(math.Max(float64(stat.max), float64(uint64(result.ResponseTime))))
+			sum = stat.sum + uint64(result.ResponseTime.Nanoseconds())
+			counter = stat.counter + 1
+
+		} else {
+			min = uint64(math.Min(float64(min), float64(uint64(result.ResponseTime))))
+			max = uint64(math.Max(float64(max), float64(uint64(result.ResponseTime))))
+			sum = uint64(result.ResponseTime)
+			counter = 1
+		}
+		avg = sum / uint64(counter)
+	}
+	return serverStats{min, max, sum, avg, errorCount, errorString, counter}
+}
+
+func getEndPoint(result madmin.AliveResult) Endpoint {
+	address := strings.Split(result.Endpoint.Host, ":")
```

Comment:
Golang has net.SplitHostPort() which is good because it can parse ipv6 addresses as well, but not so good because it returns an error when the port is not found. It is better than parsing manually anyway